clippy_lints/src/no_effect.rs

```diff
@@ -144,6 +144,7 @@ fn check_no_effect(cx: &LateContext<'_>, stmt: &Stmt<'_>) -> bool {
         }
     } else if let StmtKind::Local(local) = stmt.kind {
         if !is_lint_allowed(cx, NO_EFFECT_UNDERSCORE_BINDING, local.hir_id)
+            && !any_parent_is_automatically_derived(cx.tcx, local.hir_id)
```

Comment:
Could you move this check to the end? I believe it's expensive in comparison to the other checks. Once it's moved, we can merge this. Thank you for adding it :)

Reply:
Oh yeah, it's fine.

At the beginning, it locates in the end, but I think maybe check this in advance may avoid the left redundant check.

I'll fix this soon.

Reply:
Wait, no matter what location this check is, this line of code will always be executed if it's sure to be added.

And for whether the tested code has derived code, these several checks will be done. But if we do this check in advance, the short curcuit will block other checks, which is a cheaper cost.